src/Microsoft.Azure.Batch.SoftwareEntitlement.Server/Controllers/SoftwareEntitlementsController.cs

```diff
@@ -60,171 +69,121 @@ public class SoftwareEntitlementsController : Controller
-            {
-                Code = "EntitlementDenied",
-                Message = new ErrorMessage(message)
-            };
-
-            return StatusCode(400, error);
+            return HandleRequest(_renewRequestHandler, (requestBody, entitlementId));
         }
 
-        /// <summary>
-        /// Check to see whether the specified <c>api-version</c> is valid for software entitlements
-        /// </summary>
-        /// <param name="apiVersion">Api version from the query parameter</param>
-        /// <returns>True if it is valid, false otherwise.</returns>
-        private bool IsValidApiVersion(string apiVersion)
+        [Route("{entitlementId}")]
+        [HttpDelete]
+        [VersionRangeConstraint(minVersion: ApiVersions.ApiVersionLatest)]
+        [Produces("application/json")]
+        public IActionResult Release(
+            string entitlementId,
+            [FromQuery(Name = ApiVersions.ParameterName)] string apiVersion)
         {
-            if (string.IsNullOrEmpty(apiVersion))
-            {
-                _logger.LogDebug("No api-version specified");
-                return false;
-            }
-
-            // Check all the valid apiVersions
-            // TODO: Once this list passes three or four items, use a HashSet<string> to do the check more efficiently
+            _logger.LogInformation(
+                "[Release] Selected api-version is {ApiVersion}",
+                apiVersion);
 
-            return apiVersion.Equals(ApiVersion201705, StringComparison.Ordinal)
-                   || apiVersion.Equals(ApiVersion201709, StringComparison.Ordinal);
+            return HandleRequest(_releaseRequestHandler, entitlementId);
         }
 
-        private static bool ApiSupportsVirtualMachineId(string apiVersion)
+        private IActionResult HandleApiVersionError(string errorMessage)
         {
-            return string.Equals(apiVersion, ApiVersion201705, StringComparison.Ordinal);
+            var responseValue = new FailureResponse("EntitlementDenied", new ErrorMessage(errorMessage));
+            var response = new Response(StatusCodes.Status400BadRequest, responseValue);
+            return this.CreateActionResult(response);
         }
 
-        private static bool ApiSupportsExpiryTimestamp(string apiVersion)
+        private IActionResult HandleRequest<TRequestContext>(
+            IRequestHandler<TRequestContext> requestHandler,
+            TRequestContext requestContext)
         {
-            return string.Equals(apiVersion, ApiVersion201709, StringComparison.Ordinal);
+            try
+            {
+                var response = requestHandler.Handle(HttpContext, requestContext);
+                return this.CreateActionResult(response);
+            }
+            finally
+            {
+                if (_serverOptions.Flags.HasFlag(ServerFlags.ExitAfterRequest))
+                {
+                    _lifetime.StopApplication();
```

Comment:
George commented some time ago that this could be handled by injecting some middleware, but I never pursued that approach; would make this much simpler.

Reply:
This needs some more extensive thinking about what scenarios to support, how users should specify them, and how they should be implemented. We can do that as a separate piece of work.